e2e-polybft/e2e/consensus_test.go

```diff
@@ -823,3 +823,90 @@ func TestE2E_Deploy_Nested_Contract(t *testing.T) {
 
 	require.Equal(t, numberToPersist, parsedResponse)
 }
+
+func TestE2E_TestValidatorSetPrecompile(t *testing.T) {
+	admin, err := wallet.GenerateKey()
+	require.NoError(t, err)
+
+	validatorAddrs := []types.Address(nil)
+	adminAddr := types.Address(admin.Address())
+
+	cluster := framework.NewTestCluster(t, 4,
+		framework.WithBladeAdmin(adminAddr.String()),
+		framework.WithSecretsCallback(func(addrs []types.Address, _ *framework.TestClusterConfig) {
+			validatorAddrs = addrs
+		}))
+	defer cluster.Stop()
+
+	cluster.WaitForReady(t)
+
+	validatorKeys := make([]*wallet.Key, len(validatorAddrs))
+
+	for i, s := range cluster.Servers {
+		voterAcc, err := validatorHelper.GetAccountFromDir(s.DataDir())
+		require.NoError(t, err)
+
+		validatorKeys[i] = voterAcc.Ecdsa
+	}
+
+	txRelayer, err := txrelayer.NewTxRelayer(txrelayer.WithClient(cluster.Servers[0].JSONRPC()))
+	require.NoError(t, err)
+
+	// deploy contract
+	receipt, err := txRelayer.SendTransaction(
+		&ethgo.Transaction{
+			To:    nil,
+			Input: contractsapi.TestValidatorSetPrecompile.Bytecode,
+		},
+		admin)
+	require.NoError(t, err)
+
+	validatorSetPrecompileTestAddr := receipt.ContractAddress
+
+	hasQuorum := func() bool {
+		t.Helper()
+
+		hasQuorumFn := contractsapi.TestValidatorSetPrecompile.Abi.GetMethod("hasQuorum")
+
+		hasQuorumFnBytes, err := hasQuorumFn.Encode([]interface{}{})
+		require.NoError(t, err)
+
+		response, err := txRelayer.Call(ethgo.ZeroAddress, validatorSetPrecompileTestAddr, hasQuorumFnBytes)
+		require.NoError(t, err)
+
+		return response == "0x0000000000000000000000000000000000000000000000000000000000000001"
+	}
+
+	sendIncTx := func(validatorID int) {
```

Comment:
Maybe have a case that tests when a non validator tries to call this function. The `onlyValidator` modifier should revert the tx, since precompile will return false.

Reply:
https://github.com/Ethernal-Tech/blade/pull/118/commits/eba9e3f992d6c34c10bfdf0cf58d0b439b8a9c08

Reply:
in test there is now step where non validator sends tx which calls smart contract, but fails during the gas estimation process (because of `validatorOnly` modifier)